spec/spec.md

```diff
@@ -0,0 +1,2604 @@
+indented by the same number of spaces. To end the sub-context, a line
+must return to the indent level of the parent.
+
+Since conditional statements (`when` and `else`) may be nested, it is
+possible to create a hierarchy of indent levels, each with its own
+number of preceding spaces that must be larger than its parent's and
+consistent among all direct child statements (those that are not
+children of an even deeper conditional statement).
+
+As a concrete guide, a few consequences of these rules are summarized
+below:
+
+-   The `circuit` keyword must not be indented.
+
+-   All `module` keywords must be indented by the same number of spaces.
+
+-   In a module, all port declarations and all statements (that are not
+    children of other statements) must be indented by the same number of
+    spaces.
+
+-   The number of spaces comprising the indent level of a module is
+    specific to each module.
+
+-   The statements comprising a conditional statement's branch must be
+    indented by the same number of spaces.
+
+-   The statements of nested conditional statements establish their own,
+    deeper indent level.
+
+-   Each `when` and each `else` context may have a different number of
+    non-zero spaces in its indent level.
+
+As an example illustrating some of these points, the following is a
+legal FIRRTL circuit:
+
+    circuit Foo :
+        module Foo :
+          skip
+        module Bar :
+         input a: UInt<1>
+         output b: UInt<1>
+         when a:
+             b <= a
+         else:
+           b <= not(a)
+
+All circuits, modules, ports and statements can optionally be followed
+with the info token `@[fileinfo]` where fileinfo is a string containing
+the source file information from where it was generated. The following
+characters need to be escaped with a leading '`\`': '`\n`' (new line),
+'`\t`' (tab), '`]`' and '`\`' itself.
+
+The following example shows the info tokens included:
+
+    circuit Top : @[myfile.txt 14:8]
+       module Top : @[myfile.txt 15:2]
+         output out:UInt @[myfile.txt 16:3]
+         input b:UInt<32> @[myfile.txt 17:3]
+         input c:UInt<1> @[myfile.txt 18:3]
+         input d:UInt<16> @[myfile.txt 19:3]
+         wire a:UInt @[myfile.txt 21:8]
+         when c : @[myfile.txt 24:8]
+           a <= b @[myfile.txt 27:16]
+         else :
+           a <= d @[myfile.txt 29:17]
+         out <= add(a,a) @[myfile.txt 34:4]
+
+# FIRRTL Language Definition
+
+## Notation
+
+The concrete syntax of FIRRTL is defined in section
+[14.2](#syntax_tree).
+Productions in the syntax tree are *italicized* and keywords are written
+in `monospaced` font. The special productions , , and , indicates an
+identifier, an integer literal, and a string respectively. The notation
+$\llbracket$$\rrbracket_{...}$ is used to indicate that is repeated zero
+or more times, and the notation $\llbracket$$\rrbracket_?$ is used to
+indicate that including is optional.
+
+## Concrete Syntax Tree {#syntax_tree}
```

Comment:
This doesn't render.

Reply:
I think there are a couple of ways to handle this:

1. We can just use a fixed width region using the `ebnf` tag:

```ebnf
id        ::= letter {letter | digit}
circuit   ::= "circuit" id ":" [info] indent {module | extmodule}
module    ::= "module" id ":" [info] indent {port} {statement}
extmodule ::= "extmodule" id ":" [info] {port} [defname] {parameter}
```

2. We could use a LaTeX package that supports BNF, e.g., `\usepackage{backnaur}` and then rely on pandoc to convert this to PDF.

Orthogonally, we could move the BNF definition to be inline with each section.  E.g., when we introduce "circuit" we also have an `ebnf` block that shows its syntax.